### PR TITLE
[Text Generation] Optimize the slow `update` method in the KVCacheDecoder

### PR DESCRIPTION
**Pull request by dbogunowicz**
As reported by @mgoin and investigated by myself, the `update` method in the KVCacheDecoder is very slow.
Profiling has shown that this is due to the repeated use of the `numpy.delete` function:

image

This discussion:
https://stackoverflow.com/questions/30399534/shift-elements-in-a-numpy-array
hints that the most elegant and quite efficient replacement for `numpy.delete` would be slicing the arrays. This is the change that this PR introduces.

Short benchmarking numbers:

image

### CODE REVIEW

File: src/deepsparse/transformers/utils/decoder_kv_cache.py [hunk 1]
```diff
@@ -103,79 +103,78 @@ def update(
...
+        new_cache_array = cache_array[
+            :,
+            :,
+            bool(self._freeze_first_position) + num_non_padded_entries_to_delete :,
+            :,
+        ]
+        if self._freeze_first_position:
+            bos_entries = cache_array[:, :, :1, :]
+            new_cache_array = numpy.concatenate(
+                [bos_entries, new_cache_array], axis=self._sequence_len_axis
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/decoder_kv_cache.py:174]: could try benchmarking against idx selection

**Review by bfineran — COMMENTED** [review]
LGTM pending comment

File: src/deepsparse/transformers/utils/decoder_kv_cache.py (hunk 1 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:174]: Code used:

```python
from deepsparse.transformers.utils import DecoderKVCache
import numpy as np
import timeit

def update_session():
    session = DecoderKVCache()
    state = np.random.random((1, 32, 1000, 64))
    session.setup(session_id="dummy_id",
                  state=state,
                  num_processed_tokens=0,
                  freeze_first_position=1,
        )
    for _ in range(1000):
        session.remove_non_padded_entries(state, 500)

# run timeit on the test function
print(timeit.timeit("update_session()", setup="from __main__ import update_session", number=10))
```

File: src/deepsparse/transformers/utils/decoder_kv_cache.py [hunk 2]
```diff
@@ -103,79 +103,78 @@ def update(
             Corresponds to `input_ids.shape[1]`
         """
         self.total_num_processed_tokens += input_ids_len
-        total_cache_capacity = state[list(state.keys())[0]].shape[
+
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/utils/decoder_kv_cache.py:106]: Could we also update the docstring above on line 94; there's a typo
`taking the kv cache output of from the forward pass` --> `taking the kv cache output from the forward pass`

File: src/deepsparse/transformers/utils/decoder_kv_cache.py [hunk 3]
```diff
@@ -103,79 +103,78 @@ def update(
...
+        starting from the second oldest entry.
+        """
+        new_cache_array = cache_array[
+            :,
+            :,
+            bool(self._freeze_first_position) + num_non_padded_entries_to_delete :,
+            :,
+        ]
+        if self._freeze_first_position:
+            bos_entries = cache_array[:, :, :1, :]
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/utils/decoder_kv_cache.py:172]: What does bos stand for? beginning of sentence?

Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:172]: yes!

File: src/deepsparse/transformers/utils/decoder_kv_cache.py (hunk 2 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:106]: Good catch!

File: src/deepsparse/transformers/utils/decoder_kv_cache.py (hunk 1 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:174]: @rahul-tuli Ben probably referred to this function: https://github.com/neuralmagic/deepsparse/blob/cbc63319e3e85d2f9237559ddd0a5c7aad52735b/src/deepsparse/transformers/pipelines/text_generation.py#L422. This was how we used to (naively) do the "cache reassembly" back in the days. My code above tries to mimic this (however, avoiding the use of np.delete function). I can quickly benchmark your idea as well.

Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:174]: @rahul-tuli 
Tried this implementation:
```python
B, H, L, E = cache_array.shape
new_cache_array = numpy.empty((B, H, L - num_non_padded_entries_to_delete, E))
if self._freeze_first_position:
    new_cache_array[:,:,0,:] = cache_array[:,:,0,:]
    new_cache_array[:,:,1:,:] = cache_array[:,:,num_non_padded_entries_to_delete+1:,:]
else:
    new_cache_array[:,:,:,:] = cache_array[:,:,num_non_padded_entries_to_delete:,:]
```

roughly 1,25x slower then current implementation when freezing tokens
roughly 60x slower then current implementation when not freezing tokens

Comment (rahul-tuli) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:174]: Thanks Damian! makes sense to me now :)

**Review by rahul-tuli — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**